src/Mvc/Mvc.Core/src/ModelBinding/Metadata/DefaultBindingMetadataProvider.cs

```diff
@@ -72,6 +73,77 @@ public void CreateBindingMetadata(BindingMetadataProviderContext context)
                 context.BindingMetadata.IsBindingAllowed = bindingBehavior.Behavior != BindingBehavior.Never;
                 context.BindingMetadata.IsBindingRequired = bindingBehavior.Behavior == BindingBehavior.Required;
             }
+
+            if (GetBoundConstructor(context.Key.ModelType) is ConstructorInfo constructorInfo)
+            {
+                context.BindingMetadata.BoundConstructor = constructorInfo;
+            }
+        }
+
+        internal static ConstructorInfo GetBoundConstructor(Type type)
+        {
+            if (type.IsAbstract || type.IsValueType || type.IsInterface)
+            {
+                return null;
+            }
+
+            var constructors = type.GetConstructors();
+            if (constructors.Length == 0)
+            {
+                return null;
+            }
+
+            var parameterlessConstructor = constructors.FirstOrDefault(f => f.GetParameters().Length == 0);
+            return GetRecordTypeConstructor(type, constructors) ?? parameterlessConstructor;
+        }
+
+        private static ConstructorInfo GetRecordTypeConstructor(Type type, ConstructorInfo[] constructors)
+        {
+            if (!IsRecordType(type))
+            {
+                return null;
+            }
+
+            // For record types, we will support binding and validating the primary constructor.
+            // There isn't metadata to identify a primary constructor. We require that at most one constructor is defined on the type,
+            // and that every parameter on the constructor has a corresponding property.
```

Comment:
Considering record types only might have a non-parameterless constructor, this seems like a significant restriction. Why don't we wait for (say) Linq expression support equivalent to a `with` expression❔